src/Morphir/IR/NodeId.elm

```diff
@@ -1,48 +1,754 @@
+
+            else
+                constructNodeIdString ++ nodePathToString nodePath
+    in
     case nodeId of
-        TypeID fQName ->
-            String.concat [ "Type:", FQName.toString fQName ]
+        TypeID ( packageName, moduleName, localName ) nodePath ->
+            mapToTypeOrValue packageName moduleName localName ".type" nodePath
 
-        ValueID fQName ->
-            String.concat [ "Value:", FQName.toString fQName ]
+        ValueID ( packageName, moduleName, localName ) nodePath ->
+            mapToTypeOrValue packageName moduleName localName ".value" nodePath
 
         ModuleID ( packageName, moduleName ) ->
             String.join ":"
-                [ "Module"
-                , Path.toString Name.toTitleCase "." packageName
+                [ Path.toString Name.toTitleCase "." packageName
                 , Path.toString Name.toTitleCase "." moduleName
                 ]
+
+
+nodePathToString : NodePath -> String
+nodePathToString nodePath =
+    if List.isEmpty nodePath then
+        ""
+
+    else
+        "#"
+            ++ (nodePath
+                    |> List.map
+                        (\pathStep ->
+                            case pathStep of
+                                ChildByName name ->
+                                    Name.toCamelCase name
+
+                                ChildByIndex index ->
+                                    String.fromInt index
+                        )
+                    |> String.join ":"
+               )
+
+
+nodePathFromString : String -> NodePath
+nodePathFromString string =
+    if String.isEmpty string then
+        []
+
+    else
+        string
+            |> String.split ":"
+            |> List.map
+                (\stepString ->
+                    case String.toInt stepString of
+                        Just index ->
+                            ChildByIndex index
+
+                        Nothing ->
+                            ChildByName (Name.fromString stepString)
+                )
+
+
+returnInvalidPathError : NodePath -> Result Error value
+returnInvalidPathError pathSoFar =
+    case nodePathToString pathSoFar of
+        "" ->
+            Err <| InvalidPath "Path is invalid"
+
+        _ ->
+            Err <| InvalidPath ("Path is invalid after " ++ nodePathToString pathSoFar)
+
+
+getFromList : List NodePathStep -> List NodePathStep -> (NodePath -> NodePath -> a -> Result Error attr) -> List a -> Result Error attr
+getFromList nonEmptyPath pathSoFar recursiveFunction list =
+    case nonEmptyPath of
+        ((ChildByIndex n) as current) :: xs ->
+            list
+                |> List.Extra.getAt n
+                |> Maybe.map (recursiveFunction xs (List.append [ current ] pathSoFar))
+                |> Maybe.withDefault (returnInvalidPathError pathSoFar)
+
+        _ ->
+            returnInvalidPathError pathSoFar
+
+
+getAttribute : Package.Definition attr attr -> NodeID -> Result Error attr
```

Comment:
Could you please alias the type `attr`. Its difficult understanding what that is.